lib/modules/asset/AssetService.ts

```diff
@@ -304,213 +303,105 @@ export class AssetService {
-        await this.sdk.document.mDelete(
-          engineId,
-          InternalCollection.ASSETS,
-          assetsList
-        );
 
-        recovery.addRecovery(async () => {
-          await this.sdk.document.mCreate(
-            engineId,
-            InternalCollection.ASSETS,
-            assetsToMigrate
-          );
-        });
-
-        // Link the devices to the new assets
-        const promises: Promise<void>[] = [];
-
-        for (const asset of assetLinkedDevices) {
-          const assetId = asset.assetId;
-          for (const device of asset.linkedDevices) {
-            const deviceId = device._id;
-            const measureNames = device.measureNames;
-            promises.push(
-              ask<AskDeviceLinkAsset>("ask:device-manager:device:link-asset", {
-                assetId,
-                deviceId,
+            // ... and link this device to the asset in the new tenant
+            await ask<AskDeviceLinkAsset>(
+              "ask:device-manager:device:link-asset",
+              {
+                assetId: asset._id,
+                deviceId: device.id,
                 engineId: newEngineId,
-                measureNames: measureNames,
+                measureNames: device.measureNames,
                 user,
-              })
+              }
             );
           }
-        }
 
-        await Promise.all(promises);
-
-        recovery.addRecovery(async () => {
-          const promiseRecoveries: Promise<void>[] = [];
-
-          for (const asset of assetLinkedDevices) {
-            for (const device of asset.linkedDevices) {
-              const deviceId = device._id;
-              promiseRecoveries.push(
-                ask<AskDeviceUnlinkAsset>(
-                  "ask:device-manager:device:unlink-asset",
-                  {
-                    deviceId,
-                    user,
-                  }
-                )
-              );
-            }
-          }
-
-          await Promise.all(promiseRecoveries);
-        });
+          // Finally here, we can delete the asset in the source engine !
+          await this.sdk.document.delete(
+            engineId,
+            InternalCollection.ASSETS,
+            asset._id
+          );
 
-        // clear the groups
-        await this.sdk.document.mUpdate<AssetContent>(
-          newEngineId,
-          InternalCollection.ASSETS,
-          assetsList.map((assetId) => ({
-            _id: assetId,
-            body: {
-              groups: [],
-            },
-          }))
-        );
+          migrated++;
+        }
       } catch (error) {
-        await recovery.rollback();
         throw new BadRequestError(
           `An error occured while migrating assets: ${error}`
         );
       }
+
+      if (!migrated) {
+        throw new BadRequestError(
+          `Error occured while migrating all the assets !`
+        );
+      } else {
+        await this.sdk.collection.refresh(engineId, InternalCollection.ASSETS);
```

Comment:
You can make a for loop that iterate over each collection and refresh them, it will be cleaner

Reply:
You can do something like 

```js
const collectionsToRefresh = [
          {
            collection: InternalCollection.ASSETS,
            engine: engineId,
          },
          {
            collection: InternalCollection.DEVICES,
            engine: engineId,
          },
          {
            collection: InternalCollection.ASSETS,
            engine: newEngineId,
          },
          {
            collection: InternalCollection.DEVICES,
            engine: newEngineId,
          },
          {
            collection: InternalCollection.DEVICES,
            engine: this.config.adminIndex,
          },
        ].map(({ engine, collection }) => {
          return this.sdk.collection.refresh(engine, collection);
        });

        await Promise.all(collectionsToRefresh);
   ```
   
   The difference between the screenshot and your code is that running Promise.all will launch all request in parallel.
   
   Your code process refresh one by one, so the code runs faster using promise.all because you don't wait for the process to end before launching a new one.
   
   In javascript this is the main difference will using await vs promises.
   
   Here you don't need to await each refresh individually you can trigger them all at once.

Reply:
Understood. Great !